signer/man/ods-signerd.8.in
.TH "ods-signerd" "8" "February 2012" "OpenDNSSEC" "OpenDNSSEC ods-signerd"
.SH "NAME"
.LP
.B ods\-signerd
\- OpenDNSSEC Signer Engine daemon
.SH "SYNOPSIS"
.LP
.B ods\-signerd
.RB [ \-1 ]
.RB [ \-c 
.IR FILE ]
.RB [ \-d ] 
.RB [ \-h ] 
.RB [ \-i ] 
.RB [ \-v ] 
.RB [ \-V ] 
.P
.SH "DESCRIPTION"
.LP
ods\-signerd is part of the OpenDNSSEC software. It will keep your DNS zones
continuous signed. For more information, go to
.B http://www.opendnssec.org
and visit the Documentation page.
.P
.SH "OPTIONS"
.LP
.TP
.B \-1
Run signer engine once, then exit (for debugging purposes).
.TP
.B \-c\fI FILE
Read configuration from file, instead of using the default.
.TP
.B \-d
Run daemon in foreground.
.TP
.B \-h
Show this help.
.TP
.B \-i
Print configuration and exit (for debugging purposes).
.TP
.B \-v
Increase verbosity.
.TP
.B \-V
Show version and exit.
.P
.SH "DIAGNOSTICS"
.LP
will log all the problems via standard syslog(8).
.SH "SEE ALSO"
.LP
ods\-control(8), ods\-enforcerd(8), ods\-enforcer(8),
ods\-hsmspeed(1), ods\-hsmutil(1), ods\-kaspcheck(1),
ods\-ksmutil(1), ods\-signer(8), ods\-timing(5),
ods\-kasp(5), opendnssec(7),
.B http://www.opendnssec.org/
.SH "AUTHORS"
.LP
.B ods\-signerd
was written by NLnet Labs as part of the OpenDNSSEC project.


NEWS
$Id: NEWS 7317 2013-09-19 11:23:45Z jerry $

OpenDNSSEC 1.3.15 - 2013-09-19

* SUPPORT-58: Extend ods-signer sign <zone> with --serial <nr> so that the user
  can specify the SOA serial to use in the signed zone [OPENDNSSEC-423].
* OPENDNSSEC-428: Add option for 'ods-ksmutil key generate' to take total 
  number of zones as a parameter
* OPENDNSSEC-448: Signer Engine: Enhancements to signer debug locks.

Bugfixes:
* SUPPORT-75: Signer Engine: Fix double free crash in case of HSM connection
  error during signing [OPENDNSSEC-452].
* OPENDNSSEC-397: Change "hsmutil list" output so that the table header goes 
  to stdout not stderr
* OPENDNSSEC-438: 'ods-ksmutil key generate' and the enforcer can create 
  too many keys for <SharedKeys/> policies when KSK and ZSK use same 
  algorithm and length
* OPENDNSSEC-445: ods-ksmutil: Clean up of hsm connection handling


OpenDNSSEC 1.3.14 - 2013-05-16

* OPENDNSSEC-367: ods-ksmutil: Require user confirmation if the algorithm for
  a key is changed in a policy (as this rollover is not handled cleanly)
* OPENDNSSEC-91: Make the keytype flag required when rolling keys
* OPENDNSSEC-403: Signer Engine: new command 'ods-signer locks' that shows
  locking information (for debugging purposes).

Bugfixes:
* OPENDNSSEC-247: Signer Engine: TTL on NSEC3 was not updated on SOA
  Minimum change.
* OPENDNSSEC-396: Use TTLs from kasp when generating DNSKEY and DS records for
  output.
* OPENDNSSEC-398: The ods-ksmutil key rollover command does not work correctly
  when rolling all keys using the --policy option
* SUPPORT-40: Signer Engine: Keep occluded data in signed zone files/transfers.


OpenDNSSEC 1.3.13 - 2013-02-20

Bugfixes:
* OPENDNSSEC-388: Signer Engine: Internal serial should take into account
  the inbound serial.
* OPENDNSSEC-242: Signer Engine: Could get stuck on load signconf while
  signconf was not changed.
* Signer Engine: Fixed locking and notification on the drudge work queue,
  signals could be missed so that drudgers would stall when there was work to
  be done.


OpenDNSSEC 1.3.12 - 2012-12-03

Bugfixes:
* SUPPORT-42: ./configure fails on FreeBSD (or if ldns is not installed in a 
  directory in the default search path of the complier).
* OpenDNSSEC does not compile against ldns 1.6.16 on platforms that rely on
  the OpenDNSSEC implementation of strlcpy/cat 


OpenDNSSEC 1.3.11 - 2012-11-13

* OPENDNSSEC-330: NSEC3PARAM TTL should be set to zero.

Bugfixes:
* OPENDNSSEC-306: Cant delete zone until Enforcer made signerconf.
* OPENDNSSEC-281: Commandhandler sometimes unresponsive.
* OPENDNSSEC-299: ods-ksmutil <enter> now includes policy import
* OPENDNSSEC-300: ods-ksmutil policy purge documented with a warning
* OPENDNSSEC-338: ods-ksmutil: fix zone delete on MySQL (broken by SUPPORT-27)
* OPENDNSSEC-342: Auditor comparisons made case-insensitive
* OPENDNSSEC-345: ods-ksmutil: use ods-control to HUP the enforcerd process


OpenDNSSEC 1.3.10 - 2012-08-10

Bugfixes:
* SUPPORT-30: RRSIGs are left in the signed zone when authoritative RRsets
  become glue [OPENDNSSEC-282].
* OPENDNSSEC-261: Ldns fails to parse RR that seems syntactically correct.
  Was due to memory allocation issues. Provided better log message.
* OPENDNSSEC-285: Signer segfault for 6 or more -v options
* OPENDNSSEC-298: Only unlink existing pidfile on exit if we wrote it.
* OPENDNSSEC-303: Return if open/parse of zonelist.xml fails in ksmutil.c
  update_zones() and cmd_listzone().
* OPENDNSSEC-304: Signer Engine: Check pidfile on startup, if pidfile exists
  and corresponding process is running, then complain and exit.
* Signer seems to hang on a ods-signer command. Shutdown client explicitly
  with shutdown().
* opendnssec.spec file removed
   

OpenDNSSEC 1.3.9 - 2012-06-15

* OPENDNSSEC-277: Enforcer: Performance optimisation of database access.

Bugfixes:
* SUPPORT-27: ods-ksmutil: simplify zone delete so that it only marks keys as
  dead (rather than actually removing them). Leave the key removal to purge
  jobs.


OpenDNSSEC 1.3.8 - 2012-05-09

* OPENDNSSEC-228: Signer Engine: Make 'ods-signer update' reload signconfs
  even if zonelist has not changed. 
* OPENDNSSEC-231: Signer Engine: Allow for Classless IN-ADDR.ARPA names
  (RFC 2317).
* OPENDNSSEC-234: Enforcer: Add indexes for foreign keys in kasp DB. (sqlite
  only, MySQL already has them.)
* OPENDNSSEC-246: Signer Engine: Warn if <Audit/> is in signer configuration,
  but ods-auditor is not installed
* OPENDNSSEC-249: Enforcer: ods-ksmutil: If key export finds nothing to do
  then say so rather than display nothing which might be misinterpreted.

Bugfixes:
* OPENDNSSEC-247: Signer Engine: TTL on NSEC(3) was not updated on SOA
  Minimum change.
* OPENDNSSEC-253: Enforcer: Fix "ods-ksmutil zone delete --all"


OpenDNSSEC 1.3.7 - 2012-03-13

* OPENDNSSEC-215: Signer Engine: Always recover serial from backup,
  even if it is corrupted, preventing unnecessary serial decrementals.
* OPENDNSSEC-217: Enforcer: Tries to detect pidfile staleness, so that
  the daemon will start after a power failure.

Bugfixes:
* ods-hsmutil: Fixed a small memory leak when printing a DNSKEY.
* OPENDNSSEC-216: Signer Engine: Fix duplicate NSEC3PARAM bug.
* OPENDNSSEC-218: Signer Engine: Prevent endless loop in case the locators
  in the signer backup files and the HSM are out of sync.
* OPENDNSSEC-225: Fix problem with pid found when not existing.
* SUPPORT-21: HSM SCA 6000 in combination with OpenCryptoki can return RSA key
  material with leading zeroes. DNSSEC does not allow leading zeroes in key
  data. You are affected by this bug if your DNSKEY RDATA e.g. begins with
  "BAABA". Normal keys begin with e.g. "AwEAA". OpenDNSSEC will now sanitize
  incoming data before adding it to the DNSKEY. Do not upgrade to this version
  if you are affected by the bug. You first need to go unsigned, then do the
  upgrade, and finally sign your zone again. SoftHSM and other HSM:s will not

signer/src/ods-signerd.c
/*
 * $Id: ods-signerd.c 5227 2011-06-12 08:51:24Z jakob $
 *
 * Copyright (c) 2009 NLNet Labs. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

/**
 * OpenDNSSEC signer engine daemon.
 *
 */

#include "config.h"
#include "daemon/engine.h"

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>


#define AUTHOR_NAME "Matthijs Mekking"
#define COPYRIGHT_STR "Copyright (C) 2008-2010 NLnet Labs OpenDNSSEC"


/**
 * Prints usage.
 *
 */
static void
usage(FILE* out)
{
    fprintf(out, "Usage: %s [OPTIONS]\n", "ods-signerd");
    fprintf(out, "Start the OpenDNSSEC signer engine daemon.\n\n");
    fprintf(out, "Supported options:\n");
    fprintf(out, " -c | --config <cfgfile> Read configuration from file.\n");
    fprintf(out, " -d | --no-daemon        Do not daemonize the signer "
                 "engine.\n");
    fprintf(out, " -1 | --single-run       Run once, then exit.\n");
    fprintf(out, " -h | --help             Show this help and exit.\n");
    fprintf(out, " -i | --info             Print configuration and exit.\n");
    fprintf(out, " -v | --verbose          Increase verbosity.\n");
    fprintf(out, " -V | --version          Show version and exit.\n");
    fprintf(out, "\nBSD licensed, see LICENSE in source package for "
                 "details.\n");
    fprintf(out, "Version %s. Report bugs to <%s>.\n",
        PACKAGE_VERSION, PACKAGE_BUGREPORT);
}


/**
 * Prints version.
 *
 */
static void
version(FILE* out)
{
    fprintf(out, "%s version %s\n", PACKAGE_NAME, PACKAGE_VERSION);
    fprintf(out, "Written by %s.\n\n", AUTHOR_NAME);
    fprintf(out, "%s.  This is free software.\n", COPYRIGHT_STR);
    fprintf(out, "See source files for more license information\n");
    exit(0);
}




/**
 * Main. start engine and run it.
 *
 */
int
main(int argc, char* argv[])
{
    int c;
    int options_index = 0;
    int info = 0;
    int single_run = 0;
    int daemonize = 1;
    int cmdline_verbosity = 0;
    const char* cfgfile = ODS_SE_CFGFILE;
    static struct option long_options[] = {
        {"single-run", no_argument, 0, '1'},
        {"config", required_argument, 0, 'c'},
        {"no-daemon", no_argument, 0, 'd'},
        {"help", no_argument, 0, 'h'},
        {"info", no_argument, 0, 'i'},
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        { 0, 0, 0, 0}
    };

    /* parse the commandline */
    while ((c=getopt_long(argc, argv, "1c:dhivV",
        long_options, &options_index)) != -1) {
        switch (c) {
            case '1':
                single_run = 1;
                break;
            case 'd':
                daemonize = 0;
                break;
            case 'h':
                usage(stdout);
                exit(0);
                break;
            case 'i':
                info = 1;
                break;
            case 'v':
                cmdline_verbosity++;
                break;
            case 'V':
                version(stdout);
                exit(0);
                break;
            default:
                usage(stderr);
                exit(2);
                break;
        }
    }
    argc -= optind;
    argv += optind;
    if (argc != 0) {
        usage(stderr);
        exit(2);
    }

#ifdef ENFORCER_TIMESHIFT
    if (getenv("ENFORCER_TIMESHIFT")) {
        fprintf(stdout, "WARNING: timeshift %s detected, running once only\n",
            getenv("ENFORCER_TIMESHIFT"));
        single_run = 1;
    } else {
        fprintf(stdout, "DEBUG: timeshift mode enabled, but not set.\n");
    }
#endif /* ENFORCER_TIMESHIFT */

    /* main stuff */
    fprintf(stdout, "OpenDNSSEC signer engine version %s\n", PACKAGE_VERSION);
    engine_start(cfgfile, cmdline_verbosity, daemonize, info, single_run);

    /* done */
    return 0;
}
